src/Controller/DefaultController.php

```diff
@@ -321,13 +330,218 @@ public function rate_record($bib, $rating) {
   }
 
+  // -----------------------------------------------------------
+  // ---------------- Pickup Request-related methods -----------
+  // -----------------------------------------------------------
+  public function pickup_locations_for_patron() {
+    $returnArray = [];
+
+    $search_form = \Drupal::formBuilder()->getForm('\Drupal\arborcat\Form\ArborcatHoldsReadySearchForm');
+
+    $current_uri = \Drupal::request()->getRequestUri();
+
+    $barcode = \Drupal::request()->get('bcode');
+
+    $locationURLs = [];
+    if (isset($barcode)) {
+      $eligibleHolds = arborcat_load_patron_eligible_holds($barcode);
+      if (!isset($eligibleHolds['error'])) {
+        // grab pickup appointments to display on form
+        $scheduled_pickups = arborcat_get_scheduled_pickups($barcode);
+        // Get the patron ID from the first hold object in $eligibleHolds. NOTE - this starts at offset [1]
+        $patronId = $eligibleHolds[1]['usr'];
+        $holdLocations = [];
+        // spin through the eligible holds and get the locations
+        foreach ($eligibleHolds as $holdobj) {
+            array_push($holdLocations, $holdobj['pickup_lib']);
+        }
+        $holdLocations = array_unique($holdLocations);
+
+        $api_url = \Drupal::config('arborcat.settings')->get('api_url');
+        $guzzle = \Drupal::httpClient();
+        $locations = json_decode($guzzle->get("$api_url/locations")->getBody()->getContents());
+
+        
+        foreach ($holdLocations as $loc) {
+            $locationName = ($loc < 110) ? $locations->{$loc} : 'melcat';
+            $url = $this->createPickupURL($patronId, $barcode, $loc);
+            array_push($locationURLs, ['url'=>$url, 'loc'=>$loc, 'locname'=>$locationName]);
+        }
+      } else {
+        $locationURLs['error'] = 'Error looking up patron requests. Is this a valid barcode?';
+      }
+    }
+
+    return [
+        '#theme' => 'patron_requests_ready_locations_theme',
+        '#search_form' => $search_form,
+        '#location_urls' => $locationURLs,
+        '#barcode' => $barcode,
+        '#scheduled_pickups' => $scheduled_pickups ?? NULL
+    ];
+  }
+
+  private function createPickupURL($patronId, $barcode, $location) {
+    $html = '';
+    $pickup_requests_salt = \Drupal::config('arborcat.settings')->get('pickup_requests_salt');
+    $encryptedBarcode = md5($pickup_requests_salt . $barcode);
+    $host = \Drupal::request()->getHost();
+    $link = 'https://'. $host . '/pickuprequest/' . $patronId . '/'. $encryptedBarcode . '/' . $location;
+    return $link;
+  }
+
+  public function pickup_test() {
+    $returnval = '';
+    $barcode = \Drupal::request()->query->get('barcode');
+    $patronId = \Drupal::request()->query->get('patronid');
+    $location = \Drupal::request()->query->get('location');
+    $seeddb = \Drupal::request()->query->get('seeddb');
+    
+    $barcode =  '21621034452183';   $this->barcodeFromPatronId($patronId);
```

Comment:
Should avoid hard-coded test values in committed code, might want to just clean this up a bit, not a show stopper